src/gif.rs

```diff
@@ -105,20 +114,31 @@ impl<'a, R: 'a + Read> ImageDecoder<'a> for Decoder<R> {
             return Err(ImageError::ImageEnd);
         }
 
-        let mut buf = vec![0; self.reader.buffer_size()];
-        self.reader.read_into_buffer(&mut buf)?;
-
-        // See the comments inside `<GifFrameIterator as Iterator>::next` about
-        // the error handling of `from_raw`.
-        let image_buffer_raw = ImageBuffer::from_raw(f_width, f_height, buf).ok_or_else(
-            || ImageError::UnsupportedError("Image dimensions are too large".into())
-        )?;
+        self.reader.read_into_buffer(buf)?;
 
-        // Recover the full image
         let (width, height) = (u32::from(self.reader.width()), u32::from(self.reader.height()));
-        let image_buffer = full_image_from_frame(width, height, image_buffer_raw, left, top);
-
-        Ok(image_buffer.into_raw())
+        if (left, top) != (0, 0) || (width, height) != (f_width, f_height) {
+            // TODO: Avoid unnecessary allocation here.
```

Comment:
It would be neat to have more context here. Why is this control flow necessary, what it tries to perform and why it currently allocates.

Reply:
I added a comment explaining things better. None of the code in that block is new, it is just moved from elsewhere in the file.